cpp/test/sg/fil_test.cu

```diff
@@ -154,31 +159,33 @@ void hard_clipped_bernoulli(
     });
 }
 
+template <typename real_t>
 struct replace_some_floating_with_categorical {
   float* fid_num_cats_d;
   int num_cols;
-  __device__ float operator()(float data, int data_idx)
+  __device__ real_t operator()(real_t data, int data_idx)
   {
-    float fid_num_cats = fid_num_cats_d[data_idx % num_cols];
-    if (fid_num_cats == 0.0f) return data;
+    real_t fid_num_cats = static_cast<real_t>(fid_num_cats_d[data_idx % num_cols]);
+    if (fid_num_cats == real_t(0)) return data;
     // Transform `data` from (uniform on) [-1.0, 1.0] into [-fid_num_cats-3, fid_num_cats+3].
-    float tmp = data * (fid_num_cats + 3.0f);
+    real_t tmp = data * (fid_num_cats + real_t(3));
```

Comment:
Why not cast the entire expression and just `real_t(3)`?

Reply:
Because `data` and `fid_num_cats` are already of type `real_t`. I just want to make the cast `real_t(3)` explicit.